dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,607 @@
+	// Config defines configuration parameters that apply to each request in a claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty"`
 
-	// Selector specifies a label selector used to identify classes which
-	// may be considered equivalent to this class. In other words, this
-	// allows the adminstrator to define groups of classes which may be
-	// referred to as a single class in the claim.
+	// Requirements must be satisfied by devices. Applies to all devices of
+	// a claim when the claim references the class and only to the devices
+	// in a request when referenced there.
 	//
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
 
-	// Definition specifies the criteria to determine if a device is part
-	// of this class.
-	Definition *DeviceClassDefinition `json:"definition,omitempty"`
+// ConfigurationParameters must have one and only one field set.
+type ConfigurationParameters struct {
+	Vendor *VendorConfigurationParameters `json:"vendor,omitempty" protobuf:"bytes,1,opt,name=vendor"`
 }
 
-// DeviceClassDefinition specifies the subset of devices to consider part of
-// this class, along with device-specific configuration information for those
-// devices.
-type DeviceClassDefinition struct {
-	// Driver specifies the driver that should handle this class of devices.
-	// When a DeviceClaim uses this class, only devices published by the
-	// specified driver will be considered.
-	// +required
-	Driver string `json:driver,omitempty`
+// VendorConfigurationParameters contains configuration parameters for a driver.
+type VendorConfigurationParameters struct {
+	// DriverName is used to determine which kubelet plugin needs
+	// to be passed these configuration parameters.
+	//
+	// An admission webhook provided by the vendor could use this
+	// to decide whether it needs to validate them.
+	DriverName string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
+
+	// Parameters can contain arbitrary data. It is the
+	// responsibility of the vendor to handle validation and
+	// versioning.
+	Parameters runtime.RawExtension `json:"parameters,omitempty" protobuf:"bytes,2,opt,name=parameters"`
+}
 
-	// Constraints is a CEL expression that operates on device attributes,
-	// and must evaluate to true for a device to be considered. It will be
-	// ANDed with any Constraints field in the DeviceClaim using this class.
+// Requirement must have one and only one field set.
+type Requirement struct {
+	// Device describes a filter based on device attributes.
+	// This covers "qualitative" aspects of a device.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
+	Device *DeviceFilter `json:"device,omitempty"`
```

Comment:
If we elide the one-of, can this reduce to `Selector string` or `Constraints string`

Reply:
I was thinking of quantitative requests.

What I have simplified is the `Device` field: it's now just `DeviceSelector *string`. John, Kevin and I discussed this briefly yesterday and preferred "deviceSelector" over just "selector" because of the symmetry with other selectors (e.g. "nodeSelector").